[RISCV][TTI] Improve SiFive7 reduction cost

arcbbb opened the pull request:
The cost calculation for intrinsic llvm.vector.reduce.* factors in the costs of instructions VRED, VFRED, VMV, and VFMV.

This patch implements a target-specific cost function. If an opcode is supported in the target-specific cost function, it takes precedence over the default cost.

Co-authored-by: Elvis Wang <elvis.wang@sifive.com>

nikolaypanchenko commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -43,11 +91,23 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
   for (auto Op : OpCodes) {
+    std::optional<InstructionCost> OverrideCost =
+        GetTargetCost ? GetTargetCost(Op, VT, CostKind) : std::nullopt;
+    if (OverrideCost) {
+      Cost += *OverrideCost;
+      continue;
```
Cannot find what's a recommended way in upstream for specific cpus, but it makes more sense to either move `for (auto Op: OCodes` into `getSiFiveX280RVVCost` or call `getSiFiveX280RVVCost` under each specific case.

arcbbb replied:
I refactored `getRISCVInstructionCost`  and move `for (auto Op: OCodes)` into `getSiFiveX280RVVCost`.
Let's see if this approach makes the code look better.

sunshaoce commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -76,36 +112,62 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
+  if (CostKind == TTI::TCK_CodeSize)
+    return NumInstr;
+
+  if (ST->getProcFamily() == RISCVSubtarget::SiFive7)
```
Ask a question, can this also be used on the sifive-p670?

arcbbb replied:
No, it cannot. The latency for p670 can be observed to differ significantly from x280 in the following file: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/RISCV/RISCVSchedSiFiveP600.td.

sunshaoce replied:
Thanks for your reply.

arcbbb committed:
[RISCV][TTI] Improve SiFive7 reduction cost

The cost calculation for intrinsic llvm.vector.reduce.* factors in
the costs of instructions VRED, VFRED, VMV, and VFMV.

This patch implements a target-specific cost function.
If an opcode is supported in the target-specific cost function, it
takes precedence over the default cost.

Co-authored-by: Elvis Wang <elvis.wang@sifive.com>

arcbbb commented:
ping

wangpc-pp reviewed (commented):
I just wonder if we can make cost model bind to scheduling model.

arcbbb commented:
> I just wonder if we can make cost model bind to scheduling model.

Considering
1. The interpretation of Latency and ReleaseAtCycles in the scheduling model may not align directly with TCK_Latency and TCK_RecipThroughput in the TTI cost model.
2. In the scheduling model, the vector length is assumed to be VLMAX, whereas in the TTI cost model, we can derive the element count from fixed-length vector types.

It might be better to keep them separate for now.

wangpc-pp commented:
> > I just wonder if we can make cost model bind to scheduling model.
> 
> Considering
> 
>     1. The interpretation of Latency and ReleaseAtCycles in the scheduling model may not align directly with TCK_Latency and TCK_RecipThroughput in the TTI cost model.
> 
>     2. In the scheduling model, the vector length is assumed to be VLMAX, whereas in the TTI cost model, we can derive the element count from fixed-length vector types.
> 
> 
> It might be better to keep them separate for now.

I'm not opposed to add processor-specific cost models, but I think we should well organize the code or it may be hard to mantain as time goes on.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -76,37 +113,63 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
+    return NumInstr;
+
+  if (ST->getProcFamily() == RISCVSubtarget::SiFive7)
+    return getSiFiveX280RVVCost(OpCodes, VT, CostKind, this, TLI);
```
We don't care `CostKind` here?

arcbbb replied:
I think we care. While we treat TCK_Latency and TCK_RecipThroughput equally at the moment, it's still needed that we differentiate between them moving forward.

For simple arithmetic instructions, I thought TCK_Latency is calculated as (Latency - 1 + LMulCost/NumUnit), while TCK_RecipThroughput is calculated as (LMulCost / NumUnit) for their consuming instruction.

arcbbb replied:
My bad! So it appears that (LMulCost / NumUnit) is for independent instruction, and (Latency) is for consuming instruction.

topperc commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -76,37 +113,63 @@ RISCVTTIImpl::getRISCVInstructionCost(ArrayRef<unsigned> OpCodes, MVT VT,
+      unsigned LookUpSiFive7ReduceLatency[] = {0,  20, 27, 32, 34,
+                                               38, 40, 41, 42};
+      if (VL <= 32) {
+        Cost += LookUpSiFive7ReduceLatency[(VL + 3) >> 2];
```
Use divideCeil instead of `(VL + 3) >> 2`

arcbbb replied:
Fixed. Thanks!

arcbbb committed:
Address comments

nikolaypanchenko commented:
> I'm not opposed to add processor-specific cost models, but I think we should well organize the code or it may be hard to mantain as time goes on.

Do you have any suggestion? The only other option I can think about is each vendor will provide its own TTI, say `class VendorRISCVTTIImpl : public RISCVTTImpl`

wangpc-pp commented:
> Do you have any suggestion? The only other option I can think about is each vendor will provide its own TTI, say `class VendorRISCVTTIImpl : public RISCVTTImpl`

Currently, I have no idea. I was thinking about adapting the scheduling model.
